### PR TITLE
[move] Update syntax and usage for external resolvers in Move packages

### PR DESCRIPTION
**Pull request by tzakian**
## Description 

This updates the way external package resolvers are declared and used inside Move packages. 

In particular, external resolvers can be prefixed by `r.` so to use an external resolver binary `foo` you would declare a dependency like `A = { r.foo = "bar", <other args if you want here> }`.

This additionally adds support for resolver-specific configs in the `Move.toml` however these are treated opaquely -- any top-level item prefixed with `r.` will be allowed by the parser. 

Bottom commit are the changes + new tests, and top commit are the test updates. 
 
## Test plan 

Added new tests, as well as updated existing tests and made sure they continued to work as expected.

### CODE REVIEW

File: external-crates/move/Cargo.toml [hunk 1]
```diff
@@ -118,7 +118,7 @@ tera = "1.16.0"
 thiserror = "1.0.24"
 tiny-keccak = { version = "2.0.2", features = ["sha3"] }
 tokio = { version = "1.18.2", features = ["full"] }
-toml = "0.5.8"
+toml = {version = "0.5.8", features = ["preserve_order"]}
```
Comment (tnowacki) [inline external-crates/move/Cargo.toml:121]: Do we need preserve order any more?
Can we just "keyword" `local`, `git` and `r.<ident>`?

Comment (tzakian) [reply external-crates/move/Cargo.toml:121]: We can, although we'd then need to scan each field. But that doesn't sound too bad...

Comment (tzakian) [reply external-crates/move/Cargo.toml]: Removed! 🎉

**Review by rvantonder — APPROVED** [review]
Works for me (functionally) based on what we discussed!

Grouping dependencies inside a single dependency declaration isn't encouraged by this syntax (until we come up with something satisfying in that direction, if any), but this is totally workable. 

Thanks @tzakian!

**Review by manolisliolios — APPROVED** [review]
nice!

File: external-crates/move/crates/move-package/src/source_package/manifest_parser.rs [hunk 2]
```diff
@@ -535,11 +557,20 @@ fn parse_dep_override(tval: TV) -> Result<PM::DepOverride> {
...
+// top-level.
+fn warn_if_unknown_field_names(
+    table: &toml::map::Map<String, TV>,
+    known_names: &[&str],
+    known_prefixes: &[&str],
+) {
     let mut unknown_names = BTreeSet::new();
     for key in table.keys() {
-        if !known_names.contains(&key.as_str()) {
+        if !known_names.contains(&key.as_str())
```
Comment (amnn) [inline external-crates/move/crates/move-package/src/source_package/manifest_parser.rs:569]: I kind of expected the resolver name to show up as a key called `r` here, meaning you wouldn't need to handle prefixes specially -- was that not the case?

**Review by amnn — APPROVED** [review]
Excellent, thank you @tzakian !

**Comment by amnn** [comment]
> Grouping dependencies inside a single dependency declaration isn't encouraged by this syntax (until we come up with something satisfying in that direction, if any), but this is totally workable.

I think this is by design. Yesterday, when we were discussing the changes we would ideally make to the package resolution process, I mentioned needing to change where external resolution was handled, so that rather than waiting to hit an externally resolved dependency and then calling out to the resolver, we would gather all the externally resolved dependencies and issue a single request that would update all of them. This was to cater to the fact that the new syntax strongly encourages one dependency per declaration and yet we probably do want to avoid shelling out to the external resolver more times than we need to.

File: external-crates/move/crates/move-package/src/source_package/manifest_parser.rs (hunk 2 above)
Comment (tzakian) [reply external-crates/move/crates/move-package/src/source_package/manifest_parser.rs:569]: Just hadn't thought about/realized that. Can absolutely just look for a key called `r`.

**Pull request merged**